core/base/inc/TVirtualMutex.h

```diff
@@ -94,7 +94,7 @@ class TLockGuard {
 // TMutex m; R__LOCKGUARD(&m);
 // Warning: if program is compiled without pthread support, _REENTRANT will
 // be undefined and the macro has (silently) no effect, no locks are performed.
-#if defined (_REENTRANT) || defined (WIN32)
+#if defined (_REENTRANT) || defined (WIN32) || defined (R__FBSD)
```

Comment:
At this comment implies that test for `_REENTRANT` is an indirect way to check that the option `-pthread` is being passed to the compiler.  In FreeBSD is it essentially 'always' the case?

Reply:
in essence, yes. The choice was to build with pthread support on FreeBSD (since this is currently THE supported thread model on the platform afaik. support for some older things (N:M libkse) might still be there, but nowadays it seems all pthread to the point that _REENTRANT is not even defined anywhere in the base system (and, in particular, not set by -pthread). Which means it needs to be handled somehow manually for headers that do rely on it. What is done here is somehow a consequence of that, since llvm/interpreter does not set _REENTRANT for freebsd.

Reply:
actually, now that I had a second look, I think in the headers, if we need them to be parsable by the interpreter __FreeBSD__ needs to be defined instead of R__FBSD (as that is that llvm exports). Again, unless there is some other mechanism to set build options for interpreted code.

Reply:
> This comment seems to imply that test for `_REENTRANT` is an indirect way to check that the option `-pthread` is being passed to the compiler.

Yes, as far as I understand, this is basically a historical side-effect of older C libraries requiring that define to enable thread-safe functions. However, `_REENTRANT  is  now  obsolete` even according to the man page of `libc` (`man feature_test_macros` to be precise) on EL8. So I would argue that we should get rid of relying on it in ROOT, with FreeBSD just being the first platform that doesn't define it anymore.

> `__FreeBSD__`  needs to be defined instead of `R__FBSD`

This should be fine because as soon as we are in ROOT headers, we (should) have `RConfig.hxx` included somewhere.

> Again, unless there is some other mechanism to set build options for interpreted code...

If we really, really wanted to, we should be able to add a `-D` definition in the constructor `TCling::TCling`. However, as I argue above, I personally think we should rather move away from relying on `_REENTRANT` because it is obsolete anyway.

Reply:
> `_REENTRANT` is now obsolete

Fair enough.  We also now require (more exactly we don't offer a cmake consider that does not) build with `-pthread`.   So technically we could update (all) the code to act as if `_REENTRANT` is always defined.

Reply:
Yes, I think that's a good plan going forward, but I think that's on us as the team and should not block fixing test failures on FreeBSD right now. Do you agree?

Reply:
Of course :)